server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -180,6 +186,8 @@ public void allocate(RoutingAllocation allocation, ActionListener<Void> listener
         assert MasterService.assertMasterUpdateOrTestThread() : Thread.currentThread().getName();
         assert allocation.ignoreDisable() == false;
 
+        processNodeShutdowns(clusterService.state());
```

Comment:
Hmm I think this should be within the computation. We haven't really "processed" the new state until we've done that.